File: apps/mobile/modules/safely-masked-input/ios/MaskEngine.swift
import Foundation

enum SegmentCategory: Equatable {
    case integer
    case decimal
    case separator
}

struct StyledSegment {
    let text: String
    let category: SegmentCategory
}

struct MappedChar {
    let rawIndex: Int
    let formattedEnd: Int
}

struct MaskResult {
    let formatted: String
    let extracted: String
    let segments: [StyledSegment]
    let mapping: [MappedChar]
}

extension MaskResult {
    func cursorPosition(forRawCursor rawCursor: Int) -> Int {
        var position = 0
        for entry in mapping {
            guard entry.rawIndex < rawCursor else { break }
            position = entry.formattedEnd
        }
        return min(position, formatted.count)
    }
}

class MaskEngine {

    private static let decimalSeparators: Set<Character> = [".", ","]

    private static let ambiguous = MaskResult(formatted: "", extracted: "", segments: [], mapping: [])

    static func apply(rawInput: String, decimals: Int, decimalSeparator: String) -> MaskResult {
        let sep = decimalSeparator.isEmpty ? "." : decimalSeparator

        var integerEntries: [(rawIndex: Int, char: Character)] = []
        var decimalEntries: [(rawIndex: Int, char: Character)] = []
        var separatorRawIndex: Int?
        var separatorChar: Character?

        for (rawIndex, ch) in rawInput.enumerated() {
            if ch.isNumber {
                if separatorChar != nil {
                    if decimalEntries.count < decimals {
                        decimalEntries.append((rawIndex, ch))
                    }
                } else {
                    integerEntries.append((rawIndex, ch))
                }
            } else if decimalSeparators.contains(ch) && decimals > 0 {
                if separatorChar == nil {
                    separatorChar = ch
                    separatorRawIndex = rawIndex
                } else if ch != separatorChar {
                    return ambiguous
                }
            }
        }

        let hasDecimal = separatorChar != nil

        while integerEntries.count > 1 && integerEntries.first?.char == "0" {
            integerEntries.removeFirst()
        }

        let prependSynthetic = hasDecimal && integerEntries.isEmpty

        var formatted = ""
        var mapping: [MappedChar] = []
        var segments: [StyledSegment] = []
        var integerText = ""
        var decimalText = ""

        if prependSynthetic {
            formatted += "0"
            integerText = "0"
        }

        for entry in integerEntries {
            formatted.append(entry.char)
            integerText.append(entry.char)
            mapping.append(MappedChar(rawIndex: entry.rawIndex, formattedEnd: formatted.count))
        }
        if !integerText.isEmpty {
            segments.append(StyledSegment(text: integerText, category: .integer))
        }

        if hasDecimal {
            formatted += sep
            segments.append(StyledSegment(text: sep, category: .separator))
            if let sepRawIndex = separatorRawIndex {
                mapping.append(MappedChar(rawIndex: sepRawIndex, formattedEnd: formatted.count))
            }

            for entry in decimalEntries {
                formatted.append(entry.char)
                decimalText.append(entry.char)
                mapping.append(MappedChar(rawIndex: entry.rawIndex, formattedEnd: formatted.count))
            }
            if !decimalText.isEmpty {
                segments.append(StyledSegment(text: decimalText, category: .decimal))
            }
        }

        var extracted = integerText
        if hasDecimal {
            extracted += sep + decimalText
        }

        return MaskResult(formatted: formatted, extracted: extracted, segments: segments, mapping: mapping)
    }
}

File: apps/mobile/modules/safely-masked-input/android/src/main/java/expo/modules/safelymaskedinput/MaskEngine.kt
package expo.modules.safelymaskedinput

enum class SegmentCategory {
    INTEGER, DECIMAL, SEPARATOR
}

data class StyledSegment(val text: String, val category: SegmentCategory)

data class MaskResult(
    val formatted: String,
    val extracted: String,
    val segments: List<StyledSegment>,
    val cursorPosition: Int
)

object MaskEngine {

    private const val DECIMAL_SEPARATORS = ".,"

    private val AMBIGUOUS = MaskResult("", "", emptyList(), 0)

    fun apply(rawInput: String, decimals: Int, decimalSeparator: String): MaskResult {
        val sep = decimalSeparator.ifEmpty { "." }

        val integerDigits = StringBuilder()
        val decimalDigits = StringBuilder()
        var separatorChar: Char? = null

        for (ch in rawInput) {
            when {
                ch.isDigit() -> {
                    if (separatorChar != null) {
                        if (decimalDigits.length < decimals) decimalDigits.append(ch)
                    } else {
                        integerDigits.append(ch)
                    }
                }
                ch in DECIMAL_SEPARATORS && decimals > 0 -> {
                    if (separatorChar == null) {
                        separatorChar = ch
                    } else if (ch != separatorChar) {
                        return AMBIGUOUS
                    }
                }
            }
        }

        val hasDecimal = separatorChar != null

        var intStr = integerDigits.toString()
        while (intStr.length > 1 && intStr.startsWith("0")) {
            intStr = intStr.substring(1)
        }
        if (hasDecimal && intStr.isEmpty()) intStr = "0"

        val decStr = decimalDigits.toString()
        val formatted = StringBuilder()
        val segments = mutableListOf<StyledSegment>()

        if (intStr.isNotEmpty()) {
            formatted.append(intStr)
            segments.add(StyledSegment(intStr, SegmentCategory.INTEGER))
        }

        if (hasDecimal) {
            formatted.append(sep)
            segments.add(StyledSegment(sep, SegmentCategory.SEPARATOR))
            if (decStr.isNotEmpty()) {
                formatted.append(decStr)
                segments.add(StyledSegment(decStr, SegmentCategory.DECIMAL))
            }
        }

        var extracted = intStr
        if (hasDecimal) extracted += sep + decStr

        return MaskResult(formatted.toString(), extracted, segments, formatted.length)
    }
}
